src/main/java/com/google/api/codegen/metacode/InitCodeNode.java

```diff
@@ -369,46 +385,39 @@ private static void validateType(
-   *
-   * @param parentFieldPath The full path of the parent object of {@code typeRef}. Set to an empty
-   *     string if {@code typeRef} is a top level proto object. We need to keep track of this
-   *     because the keys in {@code sampleParamConfigMap} are full paths while {@code key} is a
-   *     simple field name.
-   */
-  private void resolveSampleParamConfigs(InitCodeContext context, String parentFieldPath) {
-    ImmutableMap<String, SampleParameterConfig> sampleParamConfigMap =
-        context.sampleParamConfigMap();
-    String fieldPath;
-    if (ROOT_KEY.equals(key)) {
-      fieldPath = "";
-    } else if (parentFieldPath.isEmpty()) {
-      fieldPath = key;
-    } else {
-      fieldPath = parentFieldPath + "." + key;
-    }
-    this.sampleParamConfig = sampleParamConfigMap.get(fieldPath);
-
-    if (sampleParamConfig != null) {
-      if (sampleParamConfig.readFromFile()) {
-        setupReadFileNode(context);
-      } else if (sampleParamConfig.isSampleArgument()) {
-        // If sample_argument_name is specified in config, set identifier to this name if the name
-        // has not been used yet; otherwise, error out.
-        Name argName = Name.anyLower(sampleParamConfig.sampleArgumentName());
-        if (!argName.equals(identifier)) {
-          Preconditions.checkArgument(
-              !context.symbolTable().contains(argName),
-              "sample_argument_name \"%s\" is already in use.",
-              sampleParamConfig.sampleArgumentName());
-          identifier = context.symbolTable().getNewSymbol(argName);
-        }
+  /** Apply {@code sampleParamConfig} to this node. */
+  private void resolveSampleParamConfig(
+      InitCodeContext context, SampleParameterConfig sampleParamConfig) {
+    if (sampleParamConfig.readFromFile()) {
+      setupReadFileNode(context, sampleParamConfig);
+    } else if (sampleParamConfig.isSampleArgument()) {
+      Name argName = Name.anyLower(sampleParamConfig.sampleArgumentName());
+      if (!argName.equals(identifier)) {
+        identifier =
+            identifierFromSampleArgumentName(context, sampleParamConfig.sampleArgumentName());
       }
-    } else {
-      // We only recursively call this method when the parent config is null since the parameter
-      // configs can never overlap.
-      children.values().forEach(child -> child.resolveSampleParamConfigs(context, fieldPath));
+    }
+  }
+
+  /** Apply {@code sampleParamConfig} to a resource path entity. */
+  private void resolveSampleParamConfig(
+      InitCodeContext context, String entityName, SampleParameterConfig sampleParamConfig) {
+    Preconditions.checkArgument(
+        !sampleParamConfig.readFromFile(),
+        "cannot configure a resource name entity to read from file.");
+    if (sampleParamConfig.sampleArgumentName() != null) {
+      Name entityIdentifier =
+          identifierFromSampleArgumentName(context, sampleParamConfig.sampleArgumentName());
```

Comment:
Oh, I see, the argument name refers to the name the user chose for the entity. (I was mistakenly thinking it was the name of the whole resource.)

And I see from the baselines that this will work also when there is no `sample_argument_name` specified. Cool.